File: Sources/ClipDeckKit/Search/TextNormalizer.swift
import Foundation

/// Normalização de texto para busca em português.
///
/// Digitar "cobranca" precisa achar "Cobrança", e "PROSPECCAO" precisa achar
/// "Prospecção". O `folding` do Foundation resolve os dois de uma vez, removendo
/// diacríticos e caixa numa passada só.
public enum TextNormalizer {
    private static let locale = Locale(identifier: "pt_BR")

    /// Versão comparável de uma string: minúscula, sem acento, sem espaço nas pontas.
    public static func fold(_ text: String) -> String {
        text
            .folding(options: [.diacriticInsensitive, .caseInsensitive], locale: locale)
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    /// Índices (em `Array(fold(text))`) onde começa cada palavra.
    ///
    /// A busca usa isso para dar bônus a matches em início de palavra: em
    /// "Follow-up pós-reunião", digitar "fpr" deve pontuar mais alto que "ollo".
    public static func wordStartOffsets(in folded: [Character]) -> Set<Int> {
        var starts: Set<Int> = []
        var previousWasSeparator = true

        for (offset, character) in folded.enumerated() {
            let isSeparator = !character.isLetter && !character.isNumber
            if !isSeparator && previousWasSeparator {
                starts.insert(offset)
            }
            previousWasSeparator = isSeparator
        }
        return starts
    }
}

File: windows/src/ClipDeck.Core/Search/TextNormalizer.cs
using System.Globalization;
using System.Text;

namespace ClipDeck.Core.Search;

/// <summary>
/// Normalização de texto para busca em português.
///
/// Digitar "cobranca" precisa achar "Cobrança", e "PROSPECCAO" precisa achar
/// "Prospecção". Decompor em NFD e descartar as marcas diacríticas resolve os
/// dois de uma vez.
/// </summary>
public static class TextNormalizer
{
    /// <summary>Versão comparável: minúscula, sem acento, sem espaço nas pontas.</summary>
    public static string Fold(string text)
    {
        if (string.IsNullOrEmpty(text)) return string.Empty;

        var decomposed = text.Normalize(NormalizationForm.FormD);
        var builder = new StringBuilder(decomposed.Length);

        foreach (var ch in decomposed)
        {
            // NonSpacingMark é onde vivem os acentos depois da decomposição.
            if (CharUnicodeInfo.GetUnicodeCategory(ch) != UnicodeCategory.NonSpacingMark)
            {
                builder.Append(ch);
            }
        }

        return builder
            .ToString()
            .Normalize(NormalizationForm.FormC)
            .ToLowerInvariant()
            .Trim();
    }

    /// <summary>
    /// Índices onde começa cada palavra no texto já normalizado.
    ///
    /// A busca usa isso para dar bônus a casamentos em início de palavra: em
    /// "Follow-up pós-reunião", digitar "fpr" deve pontuar mais que "ollo".
    /// </summary>
    public static HashSet<int> WordStartOffsets(ReadOnlySpan<char> folded)
    {
        var starts = new HashSet<int>();
        var previousWasSeparator = true;

        for (var offset = 0; offset < folded.Length; offset++)
        {
            var ch = folded[offset];
            var isSeparator = !char.IsLetter(ch) && !char.IsDigit(ch);

            if (!isSeparator && previousWasSeparator)
            {
                starts.Add(offset);
            }
            previousWasSeparator = isSeparator;
        }

        return starts;
    }
}
